### ISSUE TITLE
Call into JSON-LD via its WebIDL API instead of calling algorithms directly

### ISSUE THREAD

**Issue by jyasskin** [w3c/vc-di-eddsa#100]
In the discussion at https://github.com/w3c/json-ld-api/issues/580#issuecomment-2463356600, @gkellogg says that specs should always use the JSON-LD API instead of calling directly into JSON-LD algorithms. `data-cite="JSON-LD11-API` seems to be a good search term in this spec to find places that need fixing.

**Comment by msporny** [comment]
Hmm, I'm not sure this is necessary. We've got upwards of 17 implementations at this point and none of them seem to have been tripped up by this not calling into the JSON-LD algorithms using the WebIDL interfaces. Additionally, given that browsers don't implement the WebIDL interfaces, and don't plan to (AFAIK), I'm not seeing the benefit of making this change.

Out of curiosity, exactly how should the text change to use the WebIDL API?

/cc @gkellogg @dlongley @BigBlueHat @davidlehn to get more opinions if there is something for us to do here.

**Comment by gkellogg** [comment]
The reason for this recommendation is that the algorithms make use of options set the API (e.g., `compactArrays`, `documentLoader`, `processingMode`, ...). That's not to say that you can't use the algorithms directly, but where they call for using the value (explicit or default) of an API option, you'll need to take a separate provision for this.

As long as dependent specs are cognizant of this (e.g., using xxx for `documentLoader`) it should be fine, but the algorithms to take advantage of the envelope the API method steps provide.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2025-01-22-vcwg#section2-4) on 2025-01-22

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.4. Call into JSON-LD via its WebIDL API instead of calling algorithms directly (issue vc-di-eddsa#100)

_See github issue [vc-di-eddsa#100](https://github.com/w3c/vc-di-eddsa/issues/100)._




**Ivan Herman:** this one is on EDDSA.  
… but is likely beyond just that spec.  

**Manu Sporny:** this issue came up because Jeffrey Yaskin was doing a review on behalf of the TAG.  
… he'd spoken to Gregg Kellogg who'd suggested the use of the JSON-LD API and options.  
… that's Gregg's preference.  
… it would technically be a normative change if we changed the reference.  
… the reality is that we didn't have this reference previously, and we have 20 or so implementers.  
… there's also a request to use WebIDL which only browsers care about.  
… and we should reference that implementer feedback has shown this is implementable as described.  

**Dave Longley:** I'd also add that the WebIDL stuff in the JSON-LD API spec is optional.  
… so at most we might say you could do it the way we describe or the way that was requested.  

**Ivan Herman:** if we describe something in WebIDL it may present a false requirement that browsers MUST implement it.  
… this has caused confusion in past groups.  
… that's a danger we should avoid.  

**Manu Sporny:** yeah. true. If you invoke WebIDL, that can result in opposition and certainly causes confusion.  
… I think the response here is that we discussed it, that it lacked support from the group, because we do not want to mandate WebIDL even though that's fine.  
… and we did not want to suggest we intended to force browser implementers to implement the described WebIDL.  

> *Phillip Long:* If 20 implementers are fine without this clarification there's nothing that really must be done at this time sounds good.

> *Dave Longley:* and we have plenty of implementations already, +1.

**Ivan Herman:** great. please then mark it as closed.  

</details>

**Comment by msporny** [comment]
The WG [discussed this issue](https://github.com/w3c/vc-di-eddsa/issues/100#issuecomment-2607803930) and came to the following conclusions:

* There are 20+ implementers that did not seem to need the API call to be defined using WebIDL.
* Many of those implementers did not use the WebIDL interfaces to call the algorithms specified.
* If we mandated the use of the WebIDL interface, there was concern that we might draw objections from the browser vendors.

For these reasons, the WG decided that we did not have consensus to make this change and agreed to close the issue.

**Comment by jyasskin** [comment]
It's not clear to me that the 20+ implementations have all been tested to do the right thing with, for example, the context pinning that would be passed in using `documentLoader`. However, I agree that WebIDL isn't really the right interface language for interfaces between specifications. I think this emphasizes that the JSON-LD spec should provide a more convenient description of how to call into it by its next REC version.

**Comment by iherman** [comment]
> However, I agree that WebIDL isn't really the right interface language for interfaces between specifications. 

Indeed. We have met this issue in other specifications which were not (necessarily) aimed at exclusive browser usage. I wonder what the alternative would/could be. Any discussions on that happening somewhere?

**Comment by jyasskin** [comment]
@iherman Even browser-focused specifications have generally concluded that WebIDL isn't appropriate for calls between specifications. We generally use [Infra types](https://infra.spec.whatwg.org/#data-structures) instead. Check out https://fetch.spec.whatwg.org/#fetch-elsewhere for an example.
